README.md
EasyText Plugin for the jQuery Library 
=======================================
  
<!---Image-->
![Image.png...](http://static.jortech.com.ve/easyText/image.png)
  
EasyText is a plugin built on top of jQuery. It enhances the behaviour 
of the HTML input element of type text. It's inspired by the input in Google 
Contacts. It is still in beta. 

This plugin was built using the jQuery UI Widget Factory. For more information
[click here](http://learn.jquery.com/jquery-ui/widget-factory/)

**Current Version:** 0.1

Features
------------------------------------------------------------------------------
- Auto-resize input.

- Overall look of the input for intuitive editing.

- Store all the values the user has typed in for the session or permanently.

- Ajax capabilities to send requests to a RESTful service, enabling the 
  possibility to store the value in a database as soon as the user has typed in 
  a value  

Requirements
--------------------------------------------------------------------------------
- [jQuery](http://jquery.com/) 
- [jQuery UI](http://jqueryui.com/) 

Browser Compatibility
--------------------------------------------------------------------------------
Chrome: **```34+```**

Firefox: **```28+```**

Internet Explorer &nbsp;&nbsp;&nbsp;&nbsp;**`Not yet tested`**

Safari &nbsp;&nbsp;&nbsp;&nbsp;**`Not yet tested`**

Opera &nbsp;&nbsp;&nbsp;&nbsp;**`Not yet tested`**

Usage 
--------------------------------------------------------------------------------

<!---Code-->
```javascript
$('input[type="text"]').EasyText(attributes);
```

```attributes``` is an optional javascript object with the parameters explained 
                 below:

Attributes
--------------------------------------------------------------------------------

- **maxWidth** ```number``` *(Default: 200)* 

    The maximum size in pixels for the input.

- **minWidth** ```number``` *(Default: 80)* 

    The minimum size for the input.

- **values** ```array```

    It is an array containing all the values of the input in the same order the 
    user has typed in. The plugin will store all the values, regardless the type 
    of storage used. To delete the values, use the public method 
    *deleteStorage()*

- **maxChars** ```number``` *(Default 200)*

    The maximum amount of characters that are allowed in the input. 


- **storage** ```string``` *(Default 'array')*

    It can have 3 different values: 

    1. **array**: it will store all values from the input in an array. 

    2. **localStorage**: it will store all the values in the browser local storage 
                       so it can be accessed later

    3. **ajax**: It uses ajax capabilities to send request to a RESTful service. 
               This way,the value can be store in a database behind the scenes.
               For more information about RESTful philosophy see the wiki page 
               -> [RESTful](http://en.wikipedia.org/wiki/Representational_state_transfer)


- **ajaxConfig** 

    Single level object that contains:

    - **url**:`string` the url where to send the RESTful requests; 
             e.g., ```domain/API/```

    - **id**:`string` The identification string to form the uri; 
            e.g., ```domain/API/id```

    - **GETCallback**:`function` it process the GET request response. 
                     This function corresponds to the success parameter in the 
                     $.ajax() method described 
                     [here](https://api.jquery.com/jQuery.ajax/). If this 
                     parameter is not set, the plugin will not send a GET 
                     request when it is instantiated. The parameters are:
                     
        * **data:** ```PlainObject```javaScript object containing zero or more 
                    key-value pairs. 
                     
        * **status:** ```String``` a string describing the status of the 
                      request. More [Info](http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html).
        * **jqXHR:** ```jqXHR```  a superset of the browser's native 
                     XMLHttpRequest object. More [info](http://api.jquery.com/jQuery.ajax/#jqXHR).

        e.g.
        ```javascript
        function callback(data,status,jqXHR) {
            //code...
        }
        ```


    - **UPDATETemplate**: `string` string using John's Resig javascript 
                    [Micro-Templating](http://ejohn.org/blog/javascript-micro-templating/) 
                    function. Using this templating system, allows to transfer 
                    any information using any type of format like for example 
                    JSON,XML,YAML,etc. For refering to the input's value, in the 
                    string should be a value key that will be replace with actual
                    value. e.g. "<%= value %>"

    - **UPDATEData**: ```PlainObject``` javaScript object containing zero or 
                      more key-value pairs. 
                      It will contain all the parameters for the UPDATETemplate. 
                      For each request, the object will contain the actual value
                      of the input in the form " value: 'string' "
                    
Methods
--------------------------------------------------------------------------------
- **getNextBackValue:**
gets the previous or the next value of all the values that have been store in 
the values attribute. Has a direction parameter that acts like a flag. If it is 
true it will get the next value otherwise the previous one. Return -1 if it gets
to the extreme of the array.

- **deleteStorage**
It will delete all the saved values. It will have different behaviours depending 
on the type of storage.
    - **array**: it will delete the values store in the values attribute.
    - **localStorage**: it will delete the information store in the browser's local storage
    - **ajax**: it will send a delete request to the RESTful service.

####Method Invocation
 
Because the plugin uses the jQuery UI Widget Factory, there are two especific 
ways of calling the public methods. More
[information](http://learn.jquery.com/jquery-ui/widget-factory/widget-method-invocation/). 

#####Plugin Invocation
To invoke a method using the widget's plugin, pass the name of the method as a 
string. If the method requires arguments, pass them as additional parameters 
to the plugin.

```javascript
$('input[type="text"]').easyText('getNextBackValue','true')
```

#####Instance Invocation
Under the hoods, every instance of every widget is stored on the element using 
[jQuery.data()](http://api.jquery.com/jQuery.data/). To retrieve the instance 
object, call jQuery.data() using the 
widget's full name "easyInputEasyText".

```javascript
var jQueryObject = $('input[type="text"]').easyText();

var easyText = jQueryObject.data('easyInputEasyText');

easyText.deleteStorage();

```

Demo
--------------------------------------------------------------------------------
You can see an example in the following 
[link](http://static.jortech.com.ve/easyText/jQuery/demo/).



CHANGES.md
Changes
=======

0.1 (20-04-2014)
-------------------

- First release




testing.js
$(document).ready(function() {
    var easyText1 = $('#easyText1').easyText(),
        easyText2 = $('#easyText2').easyText({
            minWidth:40,
            maxWidth:300,
            maxChars:60}),
        easyText3 = $('#easyText3').easyText(),    
        easyText4 = $('#easyText4').easyText({
            storage: 'localStorage'
        }),
        easyText5 = $('#easyText5').easyText({
            storage:'ajax',
            ajaxConfig:{
                url:'api',
                id:'1',
                GETCallback:function(){},
                UPDATEData:{},
                UPDATETemplate:'{value:<%=value%>}'
            }
        });
   

   
   $('.easyText3 button[id="next"]').on('click',function() {
       easyText3.easyText('getNextBackValue');
   });
   $('.easyText3 button[id="back"]').on('click',function() {
       easyText3.easyText('getNextBackValue',false);
   });  
   $('.easyText3 button[id="deleteStorage"]').on('click',function() {
       easyText3.easyText('deleteStorage');
   });     
   
   $('.easyText4 button[id="next"]').on('click',function() {
       easyText4.easyText('getNextBackValue');
   });
   $('.easyText4 button[id="back"]').on('click',function() {
       easyText4.easyText('getNextBackValue',false);
   });     
   $('.easyText4 button[id="deleteStorage"]').on('click',function() {
       easyText4.easyText('deleteStorage');
   }); 
   
   $('.easyText5 button[id="deleteStorage"]').on('click',function() {
       easyText5.easyText('deleteStorage');
   }); 
        

});
   

jquery.easyText.js
/*jslint nomen: true */

/**
 * @fileOverview EasyEdit - A jQuery plugin for the input dom element of type text
 * @author Andres Jorquera
 * @version 0.2
 * @requires jQuery
 * @namespace easyInput.easyText
 * TODO Must set the new values if the storage option changes
 */

;(function ($, window, document, undefined) {
    'use strict';
    
    var NAMESPACE = 'easyInput',
        PLUG_NAME = 'easyText',
        // constant use for events handlers to get the plugin instance using
        //$(this).data(NAME_SPAC_PLUG)
        NAME_SPAC_PLUG = NAMESPACE + PLUG_NAME.charAt(0).toUpperCase() + 
                         PLUG_NAME.slice(1);

    /**
     * Description of the plugin
     * 
     */

    $.widget(NAMESPACE + '.' + PLUG_NAME, {

        options: {
            /**
             * The current value of the input element
             * @type String
             * @default ''
             */
            currentValue: '',

            /**
             * A custom array containing values for the input element
             * @type Array
             * @default []
             */
            values: [],

            /**
             * Max width in pixels for the input element.
             * @type Number
             * @default 200
             */
            maxWidth: 200,
            /**
             * Min width for the input element in pixels fo the input element.
             * @type number
             * @default 80
             */
            minWidth: 80,
            /**
             * Maximum number of characters permitted in the input.
             * @type number
             * @default 200
             */
            maxChars: 200,
            /**
             * Type of storage use for the input element.
             * @type string
             * @default 'array'
             */
            storage: 'array',
            /**
             * Configuration options for the ajax requests
             * @type object
             */
            ajaxConfig: {
                url: '/',
                id: '',
                UPDATETemplate: '',
                UPDATEData: {},
                GETCallback: null
            }
        },
        //---------------------------------------------------------------------
        //                          PRIVATE PROPERTIES
        //---------------------------------------------------------------------

        /**
         * Holder for the name of the plugin.
         * @type string
         * @private
         */
        _name: PLUG_NAME,
        /**
         * Holder for the input Jquery object.
         * @type Jquery object
         * @private
         */
        _input:null,
        /**
         * Holder for the id of the input.
         * @type string
         * @private
         */
        _id:null,
        /**
         * Holder for the current position of the current value in the
         * values attribute. It's use by the getNextBackValue public function
         * @type number
         * @private
         */
        _valuesIndex:null,
        /**
         * Holder for the class-name for the 'unChange'. Its the look of the
         * input when it hasn't been modified
         * @type String
         * @private
         */
        _classUnChanged:null,
        /**
         * Holder for the class-name for the 'empty'. The look of the input
         * when it's empty
         * @type String
         * @private
         */
        _classEmpty:null,
        /**
         * Holder for the class-name for the 'onEdit'.The look of the input
         * when it's been edited
         * @type String
         * @private
         */
        _classOnEdit: null,
        /**
         * Holder for the class-name for the "changed". The look of the input
         * when it's has been changed
         * @type String
         * @private
         */
        _classChanged:null,
        /**
         * Holder for the class-name of the input. It will be a class for the
         * input that has the plugin.
         * @type String
         * @private
         */
        _classEasyText:PLUG_NAME.toLowerCase(),

//-----------------------------------------------------------------------------
//                          PRIVATE FUNCTIONS
//-----------------------------------------------------------------------------
        /**
         * Add listeners to the input
         * @private
         */
        _setListeners:function() {
            this._input.bind({
                focus: this._focusBehaviour,
                input: this._autoGrownBehaviour,
                blur: this._blurBehaviour,
                storageChange: this._storageChange
            });
        },

        /**
         * Function that it's inspired in the getClassName function from the YUI
         * widget library.
         * reference: http://yuilibrary.com/yui/docs/api/files/widget_js_Widget.js.html#l309
         * FIXME re-think this function
         * @private
         */
        _getClassName: function(str) {
            return this._name.toLowerCase() + '-' + str;
        },
        /**
         * Sets the name for all the classes
         * @private
         */
        _setClassNames: function() {
            // set classes names
            this._classOnHover = this._getClassName('onHover');
            this._classUnChanged = this._getClassName('unChanged');
            this._classOnEdit = this._getClassName('onEdit');
            this._classChanged = this._getClassName('changed');
            this._classEmpty = this._getClassName('empty');

        },
        /**
         * Sets initial values for private variables
         * @private
         */
        _setVars: function() {
            var input = this.element;

            //save the host node to the private variable
            this._input = input;

            //creates an unique id in the element
            input.uniqueId();

            //saves the current position for the values attribute
            this._valuesIndex = 0;

            //sets the id
            this._id = this._name + '-' + input.attr('id');

            //set the maximum characters allow in the input
            input.attr('maxLength',this.options.maxChars);
        },
        /**
         * Creates a span element with the purpose for measuring the length in
         * pixels.
         * @private
         */
        _setRuler: function() {
            var rulerID    = '#' + this._name + '-ruler',
                ruler      = $(rulerID),
                input      = this._input;

            // check if a ruler already exist
            if (!ruler.length) {
                //NOTE maybe we shouldn't augment the string object
                String.prototype.visualLength = function() {
                    var ruler = $(rulerID);
                    //we need to escaped spaces
                    ruler.html(this.replace(/&/g, '&amp;').replace(/\s/g,'&nbsp;')
                          .replace(/</g, '&lt;').replace(/>/g, '&gt;'));

                    return ruler.width();
                };
                /* It will copy the same font family and size. Must be
                 * careful because if the font doesn't match it will not grown in the
                 * same way as the ruler
                 */
                ruler = $('<span id="' + rulerID.substr(1) + '"></span>');

                ruler.css({
                    fontSize:input.css('fontSize'),
                    fontFamily:input.css('font-family'),
                    fontWeight:input.css('fontWeight'),
                    position:'absolute',
                    visibility:'hidden',
                    //hides the ruler and breaks the text to avoid overflow
                    whiteSpace: 'nowrap'
                });
                $('body').after(ruler);
            }
        },
        /**
         * Prepares the configuration object for the ajax request. 
         * @private
         */
        _initializeAjax: function() {
            var ajaxConfig = this.options.ajaxConfig;
            
            //checks if there's a callback in the configuration object.
            //Do not send request if not
            if (!$.isFunction(ajaxConfig.GETCallback)) return;
            
            if (ajaxConfig.url.charAt(ajaxConfig.url.length - 1) !== '/') {
                ajaxConfig.url += '/'; 
            }
            
            if (!ajaxConfig.id) {
                ajaxConfig.id = this._id;
            }
            
            $.ajax({
                url:ajaxConfig.url + ajaxConfig.id,
                succes:ajaxConfig.GETCallback
            });
        },
        /**
         * Checks and sets the state for the input element 
         * @private
         */
        _setState:function() {
            var input         = this._input,
                text          = input.val(),
                visualLenText = text.visualLength(),
                minWidth      = this.options.minWidth,
                maxWidth      = this.options.maxWidth,
                width;

            if (visualLenText < minWidth) {
                width = minWidth;
            } else if (maxWidth < visualLenText) {
                width = maxWidth;
            } else {
                width = visualLenText;
            }

            input.width(width);

            //set the appropriate classes
            if (input.val().length === 0){
                //TODO do it in one call
                input.removeClass(this._classChanged);
                input.addClass(this._classEmpty);
            } else {
                input.addClass(this._classChanged);
            }
        },

        /**
         * Set the behaviour when the input gets clicked
         * @private
         */
        _focusBehaviour: function() {
            var that = $(this).data(NAME_SPAC_PLUG),
                input = that._input;

            input.removeClass(that._classChanged + ' ' +
                              that._classUnChanged + ' ' +
                              that._classEmpty)
                 .addClass(that._classOnEdit);
        },
        
        /**
         * Function that controls the auto-grown behaviour of the input 
         * @private
         */
        _autoGrownBehaviour: function() {
            var that     = $(this).data(NAME_SPAC_PLUG),
                input    = that._input,
                text     = input.val(),
                minWidth = that.options.minWidth,
                maxWidth = that.options.maxWidth,
                width    = text.visualLength();
                
            if (width < minWidth) {
                width = minWidth;
            } else if (width > maxWidth) {
                width = maxWidth;
            } 

            input.width(width);   
        },
        /**
         * Function that is trigger when the input loses focus. 
         * @private 
         */
        _blurBehaviour: function(){
            var that = $(this).data(NAME_SPAC_PLUG),
                input = that._input,
                value = input.val(),
                ajaxConfig = that.options.ajaxConfig,
                UPDATETemplate;

            input.removeClass(that._classOnEdit);

            //check if the field if empty. Don't store values if it is
            if (value.length === 0) {
                input.addClass(that._classEmpty);

            } else {
                
                input.addClass(that._classChanged);
                that.option('values',value);

                //TODO check this, not sure if it is the best way to do it
                that._valuesIndex = that.option('values').length - 1;

                //ajax feature
                if (that.options.storage === 'ajax') {
                    ajaxConfig.UPDATEData.value = input.val();

                    $.ajax({
                        url: ajaxConfig.url + ajaxConfig.id,
                        data: that._tmpl(ajaxConfig.UPDATETemplate,
                                         ajaxConfig.UPDATEData),
                        type: 'UPDATE'
                    });
                }
            }
            that._setState();
        },

        /**
         * Simple JavaScript Templating
         * John Resig - http://ejohn.org/ - MIT Licensed
         * ref:http://ejohn.org/blog/javascript-micro-templating/
         * @private
         */
        _tmpl : function(str, data) {
            // Figure out if we're getting a template, or if we need to
            // load the template - and be sure to cache the result.
            var fn = typeof(str) !== 'string' ?
               (function(){throw {
                   name: 'Invalid Parameter',
                   message: 'You must insert a string into the function'
               }}()):
             
              // Generate a reusable function that will serve as a template
              // generator (and which will be cached).
              new Function("obj",
                "var p=[],print=function(){p.push.apply(p,arguments);};" +
               
                // Introduce the data as local variables using with(){}
                "with(obj){p.push('" +
               
                // Convert the template into pure JavaScript
                str
                  .replace(/[\r\t\n]/g, " ")
                  .split("<%").join("\t")
                  .replace(/((^|%>)[^\t]*)'/g, "$1\r")
                  .replace(/\t=(.*?)%>/g, "',$1,'")
                  .split("\t").join("');")
                  .split("%>").join("p.push('")
                  .split("\r").join("\\'")
              + "');}return p.join('');");
           
            // Provide some basic currying to the user
            return data ? fn( data ) : fn;
        },
        /**
         * Setter function for the options of the plugin
         * @param {string} key - key string for the option value
         * @param {string} value - value for the option 
         * @private
         */
        _setOption: function(key, value) {
            var options = this.options;

            switch (key) {

            case 'values':
                //adds a new value to the array
                options.values.push(value);

                //saves the value on the local storage if it applies.
                //Add instead of replace. 
                if (options.storage === 'localStorage') {
                    localStorage.setItem(this._id,JSON.stringify(options.values));
                }

                return;
            case 'currentValue':
                this.option('values',value);
                this._input.val(value);
                break;
            }
            this._super("_setOption", key, value);

        },
        /**
         * Function that automatically runs the first time the plugin is 
         * instantiated. 
         * @private
         */
        _create: function () {
            var values,
                input,
                storage = this.options.storage;

            this._setClassNames();
            this._setVars();
            this._setListeners();


            input = this._input;
            //TODO re-think this
            input.addClass(this._classEasyText);

            //check the type of storage
            if (storage === 'localStorage') {

                //Check if there's some initial value in the localStorage
                if (localStorage[this._id]) {
                    values = JSON.parse(localStorage[this._id]);
                    this.options.values = values;

                    input.val(values[values.length - 1]);
                    this._valuesIndex = values.length - 1;
                }

            //ajax feature. Sends a GET request to the the web service.
            //TODO Not sure of this. With server side scripting we can write the
            //     values on the input.
            } else if (storage === 'ajax') {
                this._initializeAjax();
            }

            //create a ruler to measure the text in pixels
            this._setRuler();
            this._setState();
        },
        /**
         * Destroy an instantiated plugin and clean up
         * @private
         */
        destroy: function () {
            //removes any classes
            this._input.attr('class','');
            $.Widget.prototype.destroy.call(this);
        },
        
        /**
         * Deletes the storage for all the values. 
         * @public
         */
        deleteStorage: function() {
            var ajaxConfig = this.options.ajaxConfig,
                typeStorage = this.options.storage;

            if (typeStorage === 'localStorage') {
                localStorage.removeItem(this._id);

            } else if (typeStorage === 'ajax') {
                $.ajax({
                    url: ajaxConfig.url + ajaxConfig.id,
                    type: 'DELETE'
                });
            }

            this.options.values = [];
            this._input.val('');
            this._setState();
        },
        /**
         * Get the value store before or after the current value.
         * @param {boolean} direction - Flag use to set the direction of the
         *                              searched value. True is next value
         *                              and false is the previous value.
         * @public
         */
        getNextBackValue:function(direction) {
            var values = this.option('values'),
                index = this._valuesIndex;

            if (direction === undefined) {
                direction = true;
            }

            //check valid direction
            if (typeof(direction) !== 'boolean') {
                throw new Error('Invalid parameter');
            }

            //check if next or back
            index += direction ? 1 : (-1);

            //check if we are in one extreme
            if (values[index] === undefined) return -1;

            this._valuesIndex = index;
            this._input.val(values[index]);
            
            this._setState();
            
            return values[index];    
        }
    });
})(jQuery, window, document );